src/guiapi/src/display_helper.cpp

```diff
@@ -266,6 +266,108 @@ void render_icon_align(Rect16 rc, const png::Resource *res, color_t clr_back, ic
     }
 }
 
+/**
+ * @brief calculate size (in number of characters) of required rectangle, independent on font
+ * TODO use common algorithm with draw and merge those functions in one
+ *
+ * @param str                           text to be measured
+ * @param max_chars_per_line            max returned number of characters per line
+ * @param ret_numOfUTF8Chars            optional return number of characters (use nullptr if not needed)
+ * @return std::optional<size_ui16_t>   size of needed rectangle, it might be narrower than max_width
+ */
+std::optional<size_ui16_t> characters_meas_text(string_view_utf8 &str, uint16_t max_chars_per_line, uint16_t *numOfUTF8Chars) {
+    if (max_chars_per_line == 0)
+        return std::nullopt;
+
+    std::optional<int> last_space_index = std::nullopt;
+    unichar c = 0;
+
+    int chars_this_line = 0;
+    int chars_longest_line = 0;
+    int chars_tot = 0;
+    int row_no = 0;
+
+    while ((c = str.getUtf8Char()) != 0) {
+        ++chars_tot;
+        switch (c) {
+        case '\n': // new line
+            if (chars_this_line >= chars_longest_line)
+                chars_longest_line = chars_this_line;
+            ++row_no;
+            chars_this_line = 0;
+            break;
+        case ' ': // remember space position, discard multiple spaces
+
+            // erase multiple spaces
+            if (last_space_index && (*last_space_index == chars_this_line)) {
+                break; // break to avoid ++chars_this_line
+            }
+
+            // erase start space
+            if (chars_this_line == 0) {
+                break; // break to avoid ++chars_this_line
+            }
+
+            last_space_index = chars_this_line;
+
+            // DO NOT break!!!
+        default:
+            if ((chars_this_line + 1) >= max_chars_per_line) {
+                if (!last_space_index)
+                    return std::nullopt; // error, text does not fit
+
+                //  01234567890123
+                // "Hello world!!!"
+                // let's say last '!' does not fit
+                // last_space_index is 5
+                // chars_this_line is 13
+                // chars_longest_line become last_space_index (5)
+                // than we do new line
+                // and set chars_this_line 13 - 5 - 1 (1 = space)
+
+                // chars_longest_line can become last_space_index
+                chars_longest_line = std::max(chars_longest_line, *last_space_index);
+
+                // new line
+                ++row_no;
+
+                // count chars in next line
+                chars_this_line -= ((*last_space_index) + 1);
```

Comment:
I guess correct would be just `chars_this_line -= *last_space_index;`. You are correctly decrementing by 1 for the space being omitted, however, you should also increment by 1 for the character being processed right now (because the statement `++chars_this_line;` is skipped later). In your example "Hellow world!!!", you `set chars_this_line 13 - 5 - 1 (1 = space)` which results in 7, but "world!!!" is 8.